tokio/src/util/slab.rs

```diff
@@ -0,0 +1,748 @@
+            locked.head += 1;
+
+            // Increment the number of used slots
+            locked.used += 1;
+            me.used.store(locked.used, Release);
+
+            debug_assert_eq!(locked.slots.len(), locked.head);
+
+            Some((me.addr(idx), locked.slots[idx].gen_ref(me)))
+        }
+    }
+}
+
+impl<T> Page<T> {
+    /// Returns the slot index within the current page referenced by the given
+    /// address.
+    fn slot(&self, addr: Address) -> usize {
+        addr.0 - self.prev_len
+    }
+
+    /// Returns the address for the given slot
+    fn addr(&self, slot: usize) -> Address {
+        Address(slot + self.prev_len)
+    }
+}
+
+impl<T> Default for Page<T> {
+    fn default() -> Page<T> {
+        Page {
+            used: AtomicUsize::new(0),
+            slots: Mutex::new(Slots {
+                slots: Vec::new(),
+                head: 0,
+                used: 0,
+            }),
+            len: 0,
+            prev_len: 0,
+        }
+    }
+}
+
+impl<T> Page<T> {
+    /// Release a slot into the page's free list
+    fn release(&self, value: *const Value<T>) {
+        let mut locked = self.slots.lock().unwrap();
+
+        let idx = locked.index_for(value);
+        locked.slots[idx].next = locked.head as u32;
+        locked.head = idx;
+        locked.used -= 1;
+
+        self.used.store(locked.used, Release);
+    }
+}
+
+impl<T> CachedPage<T> {
+    /// Refresh the cache
+    fn refresh(&mut self, page: &Page<T>) {
+        let slots = page.slots.lock().unwrap();
+        self.slots = slots.slots.as_ptr();
+        self.init = slots.slots.len();
+    }
+
+    // Get a value by index
+    fn get(&self, idx: usize) -> &T {
+        assert!(idx < self.init);
+
+        // Safety: Pages are allocated concurrently, but are only ever
+        // **deallocated** by `Slab`. `Slab` will always have a more
+        // conservative view on the state of the slot array. Once `CachedPage`
+        // sees a slot pointer and initialized offset, it will remain valid
+        // until `compact()` is called. The `compact()` function also updates
+        // `CachedPage`.
+        unsafe {
+            let slot = self.slots.add(idx);
+            let value = slot as *const Value<T>;
+
+            &(*value).value
+        }
+    }
+}
+
+impl<T> Default for CachedPage<T> {
+    fn default() -> CachedPage<T> {
+        CachedPage {
+            slots: ptr::null(),
+            init: 0,
+        }
+    }
+}
+
+impl<T> Slots<T> {
+    /// Maps a slot pointer to an offset within the current page.
+    ///
+    /// The pointer math removes the `usize` index from the `Ref` struct,
+    /// shrinking the struct to a single pointer size.
+    ///
+    /// # Panics
+    ///
+    /// panics if the provided slot pointer is not contained by the page.
+    fn index_for(&self, slot: *const Value<T>) -> usize {
+        use std::mem;
+
+        let base = &self.slots[0] as *const _ as usize;
```

Comment:
The `&arr[0] as *const _` pattern normally makes me scared, because it is UB to access other elements in the slice through a raw pointer created like that, but it isn't a problem here.

Reply:
I added a note in the fn comments re: the pointer math being safe here.